core/src/layers/prometheus_client.rs

```diff
@@ -0,0 +1,669 @@
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::io;
+use std::sync::Arc;
+use std::task::Context;
+use std::task::Poll;
+use std::time::Instant;
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::FutureExt;
+use futures::TryFutureExt;
+use prometheus_client::metrics::counter::Counter;
+use prometheus_client::metrics::family::Family;
+use prometheus_client::metrics::histogram::Histogram;
+use prometheus_client::registry::Registry;
+
+use crate::raw::Accessor;
+use crate::raw::*;
+use crate::*;
+
+/// Add [prometheus](https://docs.rs/prometheus) for every operations.
+///
+/// # Examples
+///
+/// ```
+/// use log::debug;
+/// use log::info;
+/// use opendal::layers::PrometheusClientLayer;
+/// use opendal::services;
+/// use opendal::Operator;
+/// use opendal::Result;
+///
+/// /// Visit [`opendal::services`] for more service related config.
+/// /// Visit [`opendal::Operator`] for more operator level APIs.
+/// #[tokio::main]
+/// async fn main() -> Result<()> {
+///     // Pick a builder and configure it.
+///     let builder = services::Memory::default();
+///     let mut registry = prometheus_client::registry::Registry::default();
+///
+///     let op = Operator::new(builder)
+///         .expect("must init")
+///         .layer(PrometheusClientLayer::with_registry(&mut registry))
+///         .finish();
+///     debug!("operator: {op:?}");
+///
+///     // Write data into object test.
+///     op.write("test", "Hello, World!").await?;
+///     // Read data from object.
+///     let bs = op.read("test").await?;
+///     info!("content: {}", String::from_utf8_lossy(&bs));
+///
+///     // Get object metadata.
+///     let meta = op.stat("test").await?;
+///     info!("meta: {:?}", meta);
+///
+///     // Export prometheus metrics.
+///     let mut buf = String::new();
+///     prometheus_client::encoding::text::encode(&mut buf, &registry).unwrap();
+///     println!("## Prometheus Metrics");
+///     println!("{}", buf);
+///     Ok(())
+/// }
+/// ```
+#[derive(Debug)]
+pub struct PrometheusClientLayer {
+    metrics: PrometheusClientMetrics,
+}
+
+impl PrometheusClientLayer {
+    /// create PrometheusClientLayer while registering itself to this registry.
+    pub fn with_registry(registry: &mut Registry) -> Self {
```

Comment:
How about uses `new(registry: &mut Registry)` since we don't provide a `Default` for `PrometheusClientLayer`.